### ISSUE TITLE
DataListCheck stops updating via state change once interacted with by user

### ISSUE THREAD

**Issue by benradey** [patternfly/patternfly-react#3175]
We have a `DataList` construct that contains search results. The header row contains a 'select all' checkbox, and the result rows contain individual checkboxes.

When using the `DataListCheck` widget, we've noticed two issues:
- The 'select all' checkbox is unable to trigger the first row to appear selected
- Once the user manually checks or unchecks an individual checkbox, the 'select all' checkbox is no longer able to toggle the appearance of that checkbox.

We've verified that the underlying data structure is correct using console.log statements in the render method, and when we swap out the `DataListCheck` element for a regular `Checkbox` element instead, the behavior is correct.

Here are snippets of the UI code:
Select all checkbox:
```
<DataListCheck aria-labelledby="width-ex1-check1"
                      className="checkbox"
                      isChecked={this.state.selectAllCheckValue}
                      checked={this.state.selectAllCheckValue}
                      aria-label="controlled checkbox example"
                      id="check"
                      onChange={this.handleSelectAll}
                    />
```
Individual checkboxes:
```
                {this.state.results.map((data, key) => (
                  <DataListItemRow id="data-rows">
                  {console.log(data[Search.KEY_CHECKEDITEM], data[Search.KEY_TRANSIENTPATH])}
                      <DataListCheck aria-labelledby="width-ex3-check1"
                        className="checkbox"
                        isChecked={data[Search.KEY_CHECKEDITEM]}
                        checked={data[Search.KEY_CHECKEDITEM]}
                        aria-label="controlled checkbox example"
                        id={data[Search.KEY_TRANSIENTPATH]}
                        name={data[Search.KEY_TRANSIENTPATH]}
                        onChange={this.handleDeleteCheckboxChange}
                        key={key}
                      />

                    <DataListItemCells key={data[Search.KEY_TRANSIENTPATH]}
                      dataListCells={[
                        <DataListCell key="div-title" width={2}>
                          {this.props.userAuthenticated &&
                            <Link to={data['pant:transientPath']}>{data["jcr:title"]}</Link>}
                          {!this.props.userAuthenticated &&
                            <a href={"/" + data['pant:transientPath'] + ".preview"} target="_blank">{data["jcr:title"]}</a>}
                        </DataListCell>,
                        <DataListCell key="div-description" width={2}>
                          <span>{data["jcr:description"]}</span>
                        </DataListCell>,
                        <DataListCell key="div-transient-source">
                          <span>{data["pant:transientSource"]}</span>
                        </DataListCell>,
                        <DataListCell key="div-transient-source-name">
                          <span>{data["pant:transientSourceName"]}</span>
                        </DataListCell>,
                        <DataListCell key="div-created">
                          <span >{this.formatDate(new Date(data["pant:dateUploaded"]))}</span>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                ))}
```

And here are the handlers:
Select all:
```
  private handleSelectAll = (checked: boolean, event: FormEvent<HTMLInputElement>) => {
    console.log('selectAll start')
    console.log(this.state.results)

    this.transientPaths = []
    const newResults: any[] = []
    this.state.results.map(dataitem => {
      newResults.push(JSON.parse(JSON.stringify(dataitem))) // clones the object
      newResults[newResults.length - 1][Search.KEY_CHECKEDITEM] = checked
      if (checked) {
        this.transientPaths.push(dataitem[Search.KEY_TRANSIENTPATH])
      }
    })

    this.setState({
      deleteButtonVisible: checked,
      results: newResults,
      selectAllCheckValue: checked
    }, () => {
      console.log('selectAll finish')
      console.log(this.state.results)
    })
  }
```
Individual select:
```
  private handleDeleteCheckboxChange = (checked: boolean, event: FormEvent<HTMLInputElement>) => {
    console.log('handling check change for ', event.target['name'])
    this.transientPaths = []

    const newResults: any[] = []
    this.state.results.map(data => {
      newResults.push(JSON.parse(JSON.stringify(data))) // clones the object
      if (data[Search.KEY_TRANSIENTPATH] === event.target['name']) {
        newResults[newResults.length - 1][Search.KEY_CHECKEDITEM] = checked

        if (data[Search.KEY_CHECKEDITEM]) {
          this.transientPaths.push(data[Search.KEY_TRANSIENTPATH])
        }
      }
    })

    this.setState({
      deleteButtonVisible: this.transientPaths.length > 0,
      results: newResults,
      selectAllCheckValue: false
    })
  };
```

**Comment by rachael-phillips** [comment]
Thanks @benradey for filing this issue. We have taken a look at this, and we'll need someone from our team to investigate deeper.

**Comment by kmcfaul** [comment]
Hi @benradey 

I've been taking a look at this issue and am having some trouble reproducing the bug. Here is the codesandbox I've been testing with. https://codesandbox.io/s/data-list-check-1jjp8 

I have discovered a console warning regarding the input switching between uncontrolled and controlled when using DataListCheck, and have included a similar class MyDataListCheck with a fix which addresses it. But I am unsure that this is causing the checkboxes to stop behaving properly, because they seem to work with both DataListCheck and Checkbox.

Is the codesandbox set up in a similar way to your code? I've had to make up some dummy data so I'm wondering if that difference would be enough to prevent the bug from showing.

**Comment by benradey** [comment]
Hi @kmcfaul - I agree that your codesandbox does not display the behavior - and I was unable to tweak it to display the behavior that we had seen. It appeared to work regardless of what I threw at it.

However, I was able to show a discrepency in behavior between code in my local project and nearly identical code that I pushed to the sandbox, so that is likely a significant factor in why you had trouble recreating the issue.

For comparison:
https://codesandbox.io/s/data-list-check-my5w0
https://github.com/benradey/pantheon/blob/pf-datalistcheck/frontend/src/app/search.tsx

Something to be aware of: I *purposefully broke* the expected checkbox behavior in the handlers just to make sure that they were running and that their effects were obvious. The code _ought_ to only allow you to _check_ (and not _un_-check) when clicking on the individual row checkboxes, and the 'select all' checkbox should always _uncheck_ every row regardless of its state.

In the code sandbox, this works. In my local project, it does not - the DataListChecks behave as if no handlers are attached. However, replacing DataListCheck with Checkbox (and swapping onClick with onChange) does produce the behavior that the code dictates. So although this is not quite the original issue that I described, it does show the difference in behavior between Checkbox and DataListCheck and so is likely closely related, and is the easiest to replicate without cloning and running our original project (although you are welcome to do so).

Does that help to lead you in a workable direction?

**Comment by kmcfaul** [comment]
@benradey 

I wasn't able to build the whole project (ran into issues running podman on Mac), but I was able to clone your repo/branch and build just the frontend project to reproduce your bug.

Updating the Patternfly version fixed the issue for me. Looking at the file history, it appears this issue was fixed in #3037

The version of @patternfly/react-core you need should be 3.116.1 or later. Let me know if you are still seeing the issue.

**Comment by benradey** [comment]
@kmcfaul You are correct. Upgrading to the latest version of react-core fixes the issue on my machine as well. I'm glad this is solved!
